src/Build/BackEnd/Components/SdkResolution/SdkResolverContext.cs

```diff
@@ -12,13 +16,45 @@ namespace Microsoft.Build.BackEnd.SdkResolution
     /// </summary>
     internal sealed class SdkResolverContext : SdkResolverContextBase
     {
-        public SdkResolverContext(Framework.SdkLogger logger, string projectFilePath, string solutionPath, Version msBuildVersion, bool interactive)
+        private readonly IDictionary<string, string> _globalProperties;
+        private readonly LoggingContext _loggingContext;
+
+        public SdkResolverContext(Framework.SdkLogger logger, string projectFilePath, string solutionPath, Version msBuildVersion, bool interactive, IDictionary<string, string> globalProperties, LoggingContext loggingContext)
         {
             Logger = logger;
             ProjectFilePath = projectFilePath;
             SolutionFilePath = solutionPath;
             MSBuildVersion = msBuildVersion;
             Interactive = interactive;
+            _globalProperties = globalProperties;
+            _loggingContext = loggingContext;
+        }
+
+        public override string GetGlobalPropertyValue(string name)
+        {
+            if (_globalProperties != null && _globalProperties.TryGetValue(name, out string property))
+            {
+                return property;
+            }
+
+            return string.Empty;
+        }
+
+        public override string GetEnvironmentVariableValue(string name)
+        {
+            Traits.PropertyTrackingSetting settings = (Traits.PropertyTrackingSetting)Traits.Instance.LogPropertyTracking;
+            if ((settings & Traits.PropertyTrackingSetting.EnvironmentVariableRead) == Traits.PropertyTrackingSetting.EnvironmentVariableRead)
+            {
+                // Log this usage.
+                var args = new EnvironmentVariableReadEventArgs(
+                    name,
+                    ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("EnvironmentVariableRead", name));
+                args.BuildEventContext = _loggingContext.BuildEventContext;
+
+                _loggingContext.LogBuildEvent(args);
+            }
+
+            return Environment.GetEnvironmentVariable(name) ?? string.Empty;
```

Comment:
I'd favor not returning empty if the environment variable does not exist. In msbuildian an empty string is returned for non-existing env vars, but in C# null is returned.